### ISSUE TITLE
Supporting non-host LLVM Targets

### ISSUE THREAD

**Issue by supergrecko** [bytedeco/javacpp-presets#932]
**Supporting other Stable Targets**

As of right now (1.5.4-SNAPSHOT) the LLVM build is set to only enable the target the host we're building for<sup>(1)</sup>. Is there a specific reason we're only enabling the host target? This makes the current LLVM binary unable to cross-compile because it doesn't have the other backends.

**Supporting Experimental Targets**

As of LLVM 10.0.1 LLVM has 19<sup>(2)</sup> targets, 17<sup>(3)</sup> of which are "stable", experimental backends being AVR and ARC. If we do decide to enable all the stable targets, we should also consider providing support for the experimental backends.

We should also note that AVR has become a stable backend as of LLVM 11.0.0-rc1.

**Footnotes**

1. [In the cppbuild.sh](https://github.com/bytedeco/javacpp-presets/blob/master/llvm/cppbuild.sh#L36), the Cmake `-DLLVM_TARGETS_TO_BUILD` variable is set to `host`, or whichever target if we are cross-compiling LLVM itself. One would assume this enables the target equal to the one we're building LLVM on, but I couldn't find anything regarding `host` [in the LLVM Cmake build](https://github.com/llvm/llvm-project/blob/master/llvm/CMakeLists.txt). I do believe it is currently only enabling the targets for the host arch though, because `LLVMGetFirstTarget()` with `LLVMGetNextTarget()` only yields the platform the binary was built for.

2. List of targets https://github.com/llvm/llvm-project/tree/llvmorg-10.0.1/llvm/lib/Target

3. List of stable targets https://github.com/llvm/llvm-project/blob/llvmorg-10.0.1/llvm/CMakeLists.txt#L288

Would love to hear your thoughts about this.

**Comment by saudet** [comment]
What's the size of those binaries? If they are like 10 times larger, we'll have to see if enough people need this to enable them by default.

**Comment by supergrecko** [comment]
I build LLVM 10.0.1 for linux-x86_64 with all targets enabled and compared the library sizes. I used the binary for linux-x86_64 which the presets ship with for comparison.

The current `libLLVM-10.so` is 51 megabytes which has one target included.

For my build I ended up using clang with these cmake options: `cmake -DCMAKE_EXE_LINKER_FLAGS='-Wl' -DCMAKE_INSTALL_PREFIX=../install -DLLVM_BUILD_LLVM_DYLIB=ON -DLLVM_LINK_LLVM_DYLIB=ON -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD=all -DLLVM_ENABLE_LIBXML2=OFF -DLLVM_INCLUDE_TESTS=OFF -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD="AVR;ARC" ..`, only differences being `-DLLVM_CCACHE_BUILD=ON` removed (apart from enabling all targets).

The binary with all targets enabled has a size of 80 megabytes.

**Comment by Neiko2002** [comment]
@supergrecko can you explain your use case for multi-targets? What we usually do, is to compile everything into LLVM IR and deliver it. At the client-side, we use the LLVM runtime to compile the IR into optimized machine code and run it. Having a clang executable and multi-targets in LLVM sounds just like having a full-fledged compiler, so I wonder why do you need all this inside of Java?

**Comment by supergrecko** [comment]
I'm experimenting with an AOT compiler targeting the host arch, wasm and avr in Kotlin, requiring me to have multiple targets enabled for my LLVM build. I could definitely build my own binaries and use those alongside javacpp but I was curious as to why we were only building a single target.

Clang is not necessary for my use-case

**Comment by saudet** [comment]
@supergrecko Maybe we can limit the number of targets to the ones that are actually used by you and other users?

**Comment by saudet** [comment]
Well, it doesn't look so bad. With `-DLLVM_TARGETS_TO_BUILD=host` on Fedora 30, I get this:
```
-rw-------. 1 saudet saudet 55823166 Aug 25 12:50 llvm-linux-x86_64.jar
```
And with `-DLLVM_TARGETS_TO_BUILD=all`:
```
-rw-------. 1 saudet saudet 66008994 Aug 25 13:13 llvm-linux-x86_64.jar
```
Does this look like something that would satisfy your use case?

**Comment by supergrecko** [comment]
Sounds good to me, I would also need the AVR target, but I'll be able to wait for LLVM 11 and everything should be good.

**Comment by saudet** [comment]
Included in version 1.5.4! Thanks again for your contribution
